src/cascadia/TerminalApp/Pane.cpp

```diff
@@ -302,6 +302,400 @@ bool Pane::NavigateFocus(const FocusDirection& direction)
+// - focus: the focused pane
+// - focusIsSecondSide: If the focused pane is on the "second" side (down/right of split)
+//   relative to the branch being searched
+// - offset: the offset of the current pane
+// Return Value:
+// - A tuple of Panes, the first being the focused pane if found, and the second
+//   being the adjacent pane if it exists, and a bool that represents if the move
+//   goes out of bounds.
+Pane::FocusNeighborSearch Pane::_FindNeighborForPane(const FocusDirection& direction,
+                                                     FocusNeighborSearch searchResult,
+                                                     const bool focusIsSecondSide,
+                                                     const Pane::PanePoint offset)
+{
+    // Test if the move will go out of boundaries. E.g. if the focus is already
+    // on the second child of some pane and it attempts to move right, there
+    // can't possibly be a neighbor to be found in the first child.
+    if ((focusIsSecondSide && (direction == FocusDirection::Right || direction == FocusDirection::Down)) ||
+        (!focusIsSecondSide && (direction == FocusDirection::Left || direction == FocusDirection::Up)))
+    {
+        return searchResult;
+    }
+
+    // If we are a leaf node test if we adjacent to the focus node
+    if (_IsLeaf())
+    {
+        if (_IsAdjacent(searchResult.focus, searchResult.focusOffset, shared_from_this(), offset, direction))
+        {
+            searchResult.neighbor = shared_from_this();
+        }
+        return searchResult;
+    }
+
+    auto firstOffset = offset;
+    auto secondOffset = offset;
+    // The second child has an offset depending on the split
+    if (_splitState == SplitState::Horizontal)
+    {
+        secondOffset.y += gsl::narrow_cast<float>(_firstChild->GetRootElement().ActualHeight());
+    }
+    else
+    {
+        secondOffset.x += gsl::narrow_cast<float>(_firstChild->GetRootElement().ActualWidth());
+    }
+    auto focusNeighborSearch = _firstChild->_FindNeighborForPane(direction, searchResult, focusIsSecondSide, firstOffset);
+    if (focusNeighborSearch.neighbor)
+    {
+        return focusNeighborSearch;
+    }
+
+    return _secondChild->_FindNeighborForPane(direction, searchResult, focusIsSecondSide, secondOffset);
+}
+
+// Method Description:
+// - Searches the tree to find the currently focused pane, and if it exists, the
+//   visually adjacent pane by direction.
+// Arguments:
+// - direction: The direction to search in from the focused pane.
+// - offset: The offset, with the top-left corner being (0,0), that the current pane is relative to the root.
+// Return Value:
+// - A tuple of Panes, the first being the focused pane if found, and the second
+//   being the adjacent pane if it exists, and a bool that represents if the move
+//   goes out of bounds.
+Pane::FocusNeighborSearch Pane::_FindFocusAndNeighbor(const FocusDirection& direction, const Pane::PanePoint offset)

```

Comment:
If we view this problem from the point of the focused pane, what this function (and the other) do is basically:
* move to the parent node
* check if the _other_ child node - the one the focused pane doesn't belong to - is in the direction we want to move the pane to
  * if yes: swap panes
  * if no: continue with the parent node

Is that correct?
If my understanding is correct, I believe we can simplify this code after it has been merged.

Reply:
That is mostly a correct summary, yes. I do notice that my comment on this function is out of date so I'll update that.

Reply:
We can file this as a follow up - it's not blocking this particular PR, but yea the code probably could be improved.